Feat/generic remote tokenizer

ae86zhizhi opened the pull request:
##  Pull Request Description

This PR implements a generic remote tokenizer that communicates directly with inference engines' tokenizer APIs, ensuring perfect alignment between client-side token calculations and server-side processing.

## Related Issues 
Resolves: #1306

## Changes

### Core Implementation
- Added `RemoteTokenizer` interface and implementation.
- Implemented adapter pattern for different engines:
    - **`VLLMAdapter`**: Full implementation completed with complete tokenize/detokenize support.
    - **`SGLangAdapter`**: Basic completion endpoint support (foundation for future enhancement).
- Created a robust `HTTPClient` with:
    - Connection pooling (100 max idle, 10 per host).
    - Exponential backoff retry logic.
    - Proper resource cleanup (fixed response body leak).

### vLLM Integration (Fully Implemented)
- ✅ `/tokenize` endpoint support with all options.
- ✅ `/detokenize` endpoint support.
- ✅ Model-specific tokenizer selection.
- ✅ Special tokens handling (`add_special_tokens` parameter).
- ✅ Full compatibility with vLLM's tokenizer API.

### Error Handling
- Comprehensive error types for different scenarios.
- Graceful handling of network failures.

### Backward Compatibility
- Maintains existing `Tokenizer` interface.
- No breaking changes to current implementations.
- Opt-in via configuration.

## Testing

- [x] Unit tests for all adapters.
- [x] Integration tests with vLLM mock server.
- [x] Retry and timeout scenarios tested.
- [x] Error handling edge cases covered.
- [x] All tests passing.

## Performance Considerations

- Connection reuse reduces latency.
- Configurable timeouts prevent hanging requests.

## Usage

```go
// vLLM tokenizer - fully supported
config := RemoteTokenizerConfig{
    Engine:   "vllm",
    Endpoint: "http://vllm-server:8000",
    Model:    "meta-llama/Llama-2-7b-chat-hf",
}

tokenizer, err := NewRemoteTokenizer(config)
if err != nil {
    return err
}

// Use standard tokenizer interface
tokens, err := tokenizer.TokenizeInputText("Hello, world!")

// Or use advanced options
result, err := tokenizer.TokenizeWithOptions(ctx, TokenizeInput{
    Type:             ChatInput,
    Text:             "User message",
    AddSpecialTokens: true,
})
```

## Checklist

- [x] Code follows project style guidelines.
- [x] Tests added/updated.
- [x] Documentation updated (`README.md` included).
- [x] Linting passes.
- [x] No breaking changes.
- [x] vLLM integration fully implemented and tested.

## Future Work

- Response caching for frequent strings.
- Complete SGLang implementation.
- Additional engine support (TensorRT-LLM, etc.).

ae86zhizhi commented on pkg/utils/tokenizer/adapter_vllm.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// PrepareTokenizeRequest prepares a tokenize request for vLLM
+func (a *VLLMAdapter) PrepareTokenizeRequest(input TokenizeInput) (interface{}, error) {
```
I will use `va` instead of `a`.

autopear replied:
ok

ae86zhizhi commented on pkg/utils/tokenizer/adapter_factory.go:
```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2024 The Aibrix Team.
```
Got it.

ae86zhizhi commented on pkg/utils/tokenizer/interfaces.go:
```diff
@@ -0,0 +1,67 @@
+
+// TokenizerV2 is an alias for ExtendedTokenizer
+// Deprecated: Use ExtendedTokenizer instead
+type TokenizerV2 = ExtendedTokenizer
```
I will remove it.

Jeffwan replied:
let's remove it. this is kind of confusing since the pr is the first version

ae86zhizhi commented on pkg/utils/tokenizer/remote_tokenizer.go:
```diff
@@ -0,0 +1,220 @@
+)
+
+// RemoteTokenizerImpl implements the RemoteTokenizer interface
+type RemoteTokenizerImpl struct {
```
You’re absolutely right; change applied.

ae86zhizhi commented on pkg/utils/tokenizer/local_tiktoken.go:
```diff
@@ -29,16 +26,27 @@ import (
-func NewTiktokenTokenizer() Tokenizer {
-	return &tiktokenTokenizer{}
+// TiktokenTokenizer implements local tokenization using tiktoken
+type TiktokenTokenizer struct{}
```
You’re absolutely right; change applied.

ae86zhizhi commented on pkg/utils/tokenizer/local_characters.go:
```diff
@@ -16,13 +16,10 @@ limitations under the License.
 
-type characterTokenizer struct{}
+// CharacterTokenizer implements local tokenization by converting text to bytes
+type CharacterTokenizer struct{}
```
You’re absolutely right; change applied.

Jeffwan commented on pkg/utils/tokenizer/README.md:
```diff
@@ -0,0 +1,429 @@
+
+// Create a remote vLLM tokenizer (backward compatible)
+config := tokenizer.VLLMTokenizerConfig{
+    BaseURL: "http://vllm-service:8000",
```
em. Technically, we should be able to fetch the remote backends and forward requests. We do not necessibiliy need to feed the baseURL

Jeffwan commented on pkg/utils/tokenizer/README.md:
```diff
@@ -0,0 +1,429 @@
+    Model: "meta-llama/Llama-2-7b-hf",
+    Timeout: 30,
+}
+tok3, err := tokenizer.NewVLLMTokenizer(config)
```
it requires user to pass the vLLM engine tags to the deployment for user

ae86zhizhi replied:
I will update the readme.

Jeffwan commented on pkg/utils/tokenizer/README.md:
```diff
@@ -0,0 +1,429 @@
+│   └── adapter_sglang.go     # SGLang adapter
+│
+├── Compatibility
+│   ├── compat_vllm.go        # Legacy vLLM tokenizer
```
what's the difference between compat_vllm.go and adapter_vllm.go?

Jeffwan commented on pkg/utils/tokenizer/adapter_sglang.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// NewSGLangAdapter creates a new SGLang adapter
+func NewSGLangAdapter(model string) *SGLangAdapter {
```
this is just to demonstrate the multi-engine support? Technically, SGLang doesn't support the tokenization?

ae86zhizhi replied:
2. **SGLang** (`sglang`)
   - Limited support (no tokenize/detokenize endpoints)
   - As of 2025-07-18, SGLang Runtime's HTTP service does not expose `/tokenize` or `/detokenize` endpoints
   - Official documentation only lists endpoints like `/generate`, `/encode`, `/v1/chat/completions`, `/v1/embeddings`, etc.
   - Community has requested tokenizer endpoints (Issues #5653 and #7068) for RL/Agent frameworks, but these remain in discussion/planning phase
   - **Workarounds if you need tokenization with SGLang**:
     - **Client-side tokenization**: Use HuggingFace Tokenizer or similar tools to convert text to `input_ids` before calling SGLang endpoints
     - **Skip server tokenization**: Start SGLang with `--skip-tokenizer-init` flag and pass `input_ids` arrays directly in requests
   - The adapter is included as a placeholder for future implementation when/if SGLang adds tokenizer endpoints

Jeffwan replied:
sounds good

Jeffwan commented on pkg/utils/tokenizer/factory.go:
```diff
@@ -0,0 +1,27 @@
+limitations under the License.
+*/
+
+package tokenizer
```
RemoteTokenizer is not included here?

ae86zhizhi replied:
I’ve already refactored the package following @autopear 's suggestion.

Jeffwan reviewed (commented):
seems the change scope is very clear and extensible. Let's quickly polish this one and merge it

varungup90 commented on pkg/utils/tokenizer/compat_vllm.go:
```diff
@@ -0,0 +1,113 @@
+
+// NewVLLMTokenizer creates a new vLLM tokenizer instance
+// Deprecated: Use NewRemoteTokenizer with RemoteTokenizerConfig instead
+func NewVLLMTokenizer(config VLLMTokenizerConfig) (*VLLMTokenizer, error) {
```
I did not get the part, why is there a need for NewVLLMTokenizer and not directly use adapter_vllm via remote tokenizer?

autopear reviewed (commented):
cmt

ae86zhizhi commented:
This PR has been updated with a comprehensive
refactoring of the tokenizer package to improve code
 quality, simplify the API, and enhance
maintainability. All commits now include DCO
sign-off.

Summary of Changes

1. API Simplification
- Made internal implementation types unexported
(extendedTokenizer, remoteTokenizer, engineAdapter,
etc.) to minimize the public API surface
- Removed deprecated TokenizerV2 interface alias -
users should use ExtendedTokenizer directly
- Removed the VLLM compatibility layer
(compat_vllm.go) - users should migrate to
RemoteTokenizer with Engine="vllm"
- Simplified the factory pattern by removing
unnecessary abstraction layers

2. Code Quality Improvements
- Updated all files with 2025 copyright year
- Fixed receiver naming conventions (e.g.,
VLLMAdapter from 'a' to 'va')
- Addressed all linting issues (removed extra blank
lines, added proper error handling)
- Simplified exponential backoff calculation without
 changing behavior

3. Enhanced Features
- Improved HTTP status code handling with precise
retry logic and Retry-After header support
- Optimized health checks to use empty strings,
minimizing server processing overhead
- Better error handling for JSON encoding operations
 in tests

4. Testing and Documentation
- Increased test coverage from 51.8% to 66.7%
- Removed deprecated test files for removed
components
- Completely rewrote README.md to reflect the new
simplified API
- Added migration guide for users upgrading from the
 old API
- Clarified that tokenizer.go is the main entry
point, not deprecated

Breaking Changes

- TokenizerV2 type alias removed (use
ExtendedTokenizer)
- VLLM compatibility layer removed (use
RemoteTokenizer with Engine="vllm")
- Many previously exported types are now unexported
(use type assertions for advanced features)

Benefits

- Cleaner API: Minimal public surface following Go
best practices
- Better Maintainability: Less code, clearer
structure
- Improved Performance: Optimized health checks and
retry logic
- Enhanced Documentation: Clear migration path and
usage examples

The refactoring maintains full backward
compatibility for the core Tokenizer interface while
 significantly improving the internal
implementation.

ae86zhizhi commented:
This PR has been updated with a comprehensive refactoring of the tokenizer package to improve code quality, simplify the API, and enhance maintainability. All commits now include DCO sign-off.

### Summary of Changes

**1. API Simplification**
*   Made internal implementation types unexported (extendedTokenizer, remoteTokenizer, engineAdapter, etc.) to minimize the public API surface.
*   Removed deprecated TokenizerV2 interface alias - users should use ExtendedTokenizer directly.
*   Removed the VLLM compatibility layer (compat_vllm.go) - users should migrate to RemoteTokenizer with Engine="vllm".
*   Simplified the factory pattern by removing unnecessary abstraction layers.

**2. Code Quality Improvements**
*   Updated all files with 2025 copyright year.
*   Fixed receiver naming conventions (e.g., VLLMAdapter from 'a' to 'va').
*   Addressed all linting issues (removed extra blank lines, added proper error handling).
*   Simplified exponential backoff calculation without changing behavior.

**3. Enhanced Features**
*   Improved HTTP status code handling with precise retry logic and Retry-After header support.
*   Optimized health checks to use empty strings, minimizing server processing overhead.
*   Better error handling for JSON encoding operations in tests.

**4. Testing and Documentation**
*   Increased test coverage from 51.8% to 66.7%.
*   Removed deprecated test files for removed components.
*   Completely rewrote README.md to reflect the new simplified API.
*   Added migration guide for users upgrading from the old API.
*   Clarified that tokenizer.go is the main entry point, not deprecated.

### Breaking Changes

*   TokenizerV2 type alias removed (use ExtendedTokenizer).
*   VLLM compatibility layer removed (use RemoteTokenizer with Engine="vllm").
*   Many previously exported types are now unexported (use type assertions for advanced features).

### Benefits

*   **Cleaner API:** Minimal public surface following Go best practices.
*   **Better Maintainability:** Less code, clearer structure.
*   **Improved Performance:** Optimized health checks and retry logic.
*   **Enhanced Documentation:** Clear migration path and usage examples.

The refactoring maintains full backward compatibility for the core Tokenizer interface while significantly improving the internal implementation.

autopear commented on pkg/utils/tokenizer/adapter_vllm.go:
```diff
@@ -0,0 +1,130 @@
+// ParseTokenizeResponse parses a vLLM tokenize response
+func (a *VLLMAdapter) ParseTokenizeResponse(data []byte) (*TokenizeResult, error) {
+	var resp VLLMTokenizeResponse
+	if err := json.Unmarshal(data, &resp); err != nil {
```
Whether to ignore unknown fields does not matter here. Based on 

```go
type vllmTokenizeResponse struct {
	Count       int      `json:"count"`
	MaxModelLen int      `json:"max_model_len"`
	Tokens      []int    `json:"tokens"`
	TokenStrs   []string `json:"token_strs,omitempty"`
}
```

I'm assuming if `count`, `max_model_len` or `tokens` is missing, `json.Unmarshal` will fail?  Also is `token_strs` optional, and missing this field does not throw an error?

If this is the case, add a comment above `json.Unmarshal` describing the behavior, which fields are required and which are optional.

autopear commented on pkg/utils/tokenizer/interfaces.go:
```diff
@@ -0,0 +1,63 @@
+	TokenizeInputText(string) ([]byte, error)
+}
+
+// extendedTokenizer represents an extended tokenizer interface with advanced features
```
Extend the comment for some examples of `advanced features`. Also I'm not sure if you need an `extendedTokenizer` between `Tokenizer` and `remoteTokenizer`. Why not subclass `remoteTokenizer` directly from `Tokenizer`?

autopear commented on pkg/utils/tokenizer/local_tiktoken.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2024 The Aibrix Team.
+Copyright 2025 The Aibrix Team.
```
don't change this for existing files, just use 2025 for new files

autopear commented on pkg/plugins/gateway/algorithms/prefix_cache.go:
```diff
@@ -54,13 +54,12 @@ type prefixCacheRouter struct {
-	var tokenizerObj tokenizer.Tokenizer
-	// TODO: refactor initilization
+	// Create tokenizer using factory method
 	// supported tokenizers: ["character", "tiktoken"]
```
Still just these 2 tokenizers are supported? What about remote/vllm?

autopear commented on pkg/plugins/gateway/algorithms/prefix_cache.go:
```diff
@@ -54,13 +54,12 @@ type prefixCacheRouter struct {
-		tokenizerObj = tokenizer.NewTiktokenTokenizer()
-	} else {
-		tokenizerObj = tokenizer.NewCharacterTokenizer()
+	tokenizerObj, err := tokenizer.NewTokenizer(tokenizerType, nil)
```
If remote tokenizers are supported, should the 2nd argument be non-`nil`?

autopear reviewed (changes requested):
See my comments for the revision

autopear reviewed (approved):
LGTM
